### DISCUSSION TITLE
[gym] What is the best way to add an observation ?

### DISCUSSION THREAD

**Discussion by mwulfman** [Wandercraft/jiminy#828] [Q&A] [answered]
I was wondering, if I wanted to add a trivial observation to an environment (Digit for instance), for instance a velocity commands vector, what is, in your opinion, the best way to do so? 
The observation is mainly constant but it can still change during a simulation. We should also be able to randomize it.
From the different examples in Jiminy, I was thinking of taking advantage of the `pipeline` features. Using an ObserverBlock seems to meet the criteria perfectly but seems almost overly complex for this?
Curious to know your thought.

**Comment by duburcqa** [comment, marked as answer]
It is a good question. My objective is to make everything composable in the end. Even though we are not quite there yet, I think it is in much better shape today than last year. 

Pipeline design is especially useful when you want to:
   * be able to add/remove layers at any time to test different configurations
   * have a clear separation of the input / output and internal state of each block (incl. not having access to simulator and robot state)
I see the base environment as a thin abstraction layer between the low-level motor commands and the raw sensor data, eventually with some intrinsic reward and termination conditions at must always apply. In this perspective, it is never supposed to change. In your case, I think it makes sense to consider that the task itself is independent of the environment, as part of a higher level abstraction layer. One other option would be to derive the base environment, and use this derived environment in place of the base one in your pipeline. It is more a matter of taste in your case, since being able to unwrap the task layer to get access to the base environment is not really interesting.

Still, I would say using an observer block makes slightly more sense. If it is overly complex, may be it is worth adding some extra layer deriving from it, like `TaskBlock` that would expose a more targeted API for this specific use case. 

In general, one thing that I don't like much is the inability to properly detect the first simulation step ever (which happens during env reset, after simulation reset). In some cases `_setup` is fine because having access to the simulation step is not necessary, but it is, then there is no other option than a systematic runtime check in `refresh_observation`, which is far from satisfactory. Although, this issue is more general and not limited to pipeline design, it slightly breaks the abstraction with the underlying physics engine under the hood.
